NitroxClient/MonoBehaviours/Gui/MainMenu/ServerJoin/JoinServerBackend.cs

```diff
@@ -0,0 +1,169 @@
+    }
+
+    private static void SessionConnectionStateChangedHandler(IMultiplayerSessionConnectionState state)
+    {
+        switch (state.CurrentStage)
+        {
+            case MultiplayerSessionConnectionStage.ESTABLISHING_SERVER_POLICY:
+                Log.Info("Requesting session policy info");
+                Log.InGame(Language.main.Get("Nitrox_RequestingSessionPolicy"));
+                break;
+
+            case MultiplayerSessionConnectionStage.AWAITING_RESERVATION_CREDENTIALS:
+                Color.RGBToHSV(activePlayerPreference.PreferredColor(), out float hue, out float saturation, out float brightness); // HSV => Hue Saturation Value, HSB => Hue Saturation Brightness
+                MainMenuJoinServerPanel.Instance.UpdatePanelValues(serverIp, activePlayerPreference.PlayerName, new Vector3(hue, saturation, brightness));
+
+                if (multiplayerSession.SessionPolicy.RequiresServerPassword)
+                {
+                    Log.Info("Waiting for server password input");
+                    Log.InGame(Language.main.Get("Nitrox_WaitingPassword"));
+                    MainMenuEnterPasswordPanel.ResetLastEnteredPassword();
+                    MainMenuRightSide.main.OpenGroup(MainMenuEnterPasswordPanel.NAME);
+                    break;
+                }
+
+                Log.Info("Waiting for user input");
+                Log.InGame(Language.main.Get("Nitrox_WaitingUserInput"));
+                MainMenuRightSide.main.OpenGroup(MainMenuJoinServerPanel.NAME);
+                break;
+
+            case MultiplayerSessionConnectionStage.SESSION_RESERVED:
+                Log.Info("Launching game");
+                Log.InGame(Language.main.Get("Nitrox_LaunchGame"));
+                multiplayerSession.ConnectionStateChanged -= SessionConnectionStateChangedHandler;
+                preferencesManager.Save();
+
+#pragma warning disable CS0618 // God Damn it UWE...
+                Multiplayer.SubnauticaLoadingStarted();
+                IEnumerator startNewGame = uGUI_MainMenu.main.StartNewGame(GameMode.Survival);
+#pragma warning restore CS0618 // God damn it UWE...
+                UWE.CoroutineHost.StartCoroutine(startNewGame);
+                LoadingScreenVersionText.Initialize();
+
+                break;
+
+            case MultiplayerSessionConnectionStage.SESSION_RESERVATION_REJECTED:
+                Log.Info("Reservation rejected");
+                Log.InGame(Language.main.Get("Nitrox_RejectedSessionPolicy"));
+
+                MultiplayerSessionReservationState reservationState = multiplayerSession.Reservation.ReservationState;
+
+                string reservationRejectionNotification = reservationState.Describe();
+
+                MainMenuNotificationPanel.ShowMessage(reservationRejectionNotification, null, () =>
+                {
+                    multiplayerSession.Disconnect();
+                    multiplayerSession.ConnectAsync(serverIp, serverPort);
```

Comment:
To be true, I just copied this part from the original. I believe it's so that you can immediately change input like player name or password if it was the reason for the rejection.